--- a/atf-20220606-637ba581b/plat/mediatek/mt7981/drivers/pll/pll.c
+++ b/atf-20220606-637ba581b/plat/mediatek/mt7981/drivers/pll/pll.c
@@ -70,8 +70,14 @@ unsigned int mtk_get_cpu_freq()
 
 void mtk_pll_init(int skip_dcm_setting)
 {
-
-	/* Power on PLL */
+	uint32_t readreg;
+	mmio_clrbits_32(ARMPLL_CON1, ARMPLL_CON1_ALL0); // clear all armpll_con1
+	readreg = mmio_read_32(ARMPLL_CON1); //read con1
+	NOTICE("CON1 should 0 actual %X",readreg); //print con1
+	mmio_write_32(ARMPLL_CON1, 0x52000000);
+	mmio_clrbits_32(ARMPLL_CON0 , DIVIDE_RATIO_BIT4);
+
+
 	mmio_setbits_32(ARMPLL_PWR_CON0, CON0_PWR_ON);
 	mmio_setbits_32(NET2PLL_PWR_CON0, CON0_PWR_ON);
 	mmio_setbits_32(MMPLL_PWR_CON0, CON0_PWR_ON);
@@ -92,13 +98,18 @@ void mtk_pll_init(int skip_dcm_setting)
 	mmio_clrbits_32(NET1PLL1_PWR_CON0, CON0_ISO_EN);
 	mmio_clrbits_32(APLL2_PWR_CON0, CON0_ISO_EN);
 	mmio_clrbits_32(MPLL_PWR_CON0, CON0_ISO_EN);
+
+
+	mmio_clrbits_32(ARMPLL_CON0 , DIVIDE_RATIO_BIT4);  //can clear - depand on armpll_con1
 
 	/* Set PLL frequency */
-	mmio_write_32(ARMPLL_CON1, 0x82000000); /* 1.3G */
 
-	//mmio_setbits_32(ARMPLL_CON0, 0x124); /* divider for 650M */
+	readreg = mmio_read_32(ARMPLL_CON1); //read con1
+	NOTICE("CON1 should 0 actual %X",readreg); //print con1
+	mmio_write_32(ARMPLL_CON1, 0x52000000);
+	mmio_setbits_32(ARMPLL_CON0, 0x104);
 
-	mmio_setbits_32(ARMPLL_CON0, 0x114); /* divider for 1.3G */
+
 
 	mmio_setbits_32(NET2PLL_CON0, 0x114);
 	mmio_setbits_32(MMPLL_CON0, 0x124);
--- a/atf-20220606-637ba581b/plat/mediatek/mt7981/drivers/pll/pll.h
+++ b/atf-20220606-637ba581b/plat/mediatek/mt7981/drivers/pll/pll.h
@@ -64,7 +64,8 @@
 #define CON0_POST_DIV(x)	((x) << 1)
 #define CON1_PCW_CHG		BIT(31)
 #define UNIV_48M_EN		BIT(27)
-
+#define DIVIDE_RATIO_BIT4	0x10
+#define ARMPLL_CON1_ALL0        0xFF000000
 #define SCP_AXICK_DCM_DIS_EN	BIT(0)
 #define SCP_AXICK_26M_SEL_EN	BIT(4)

--- a/atf-20231013-0ea67d76a/plat/mediatek/mt7981/drivers/pll/pll.c
+++ b/atf-20231013-0ea67d76a/plat/mediatek/mt7981/drivers/pll/pll.c
@@ -70,8 +70,14 @@ unsigned int mtk_get_cpu_freq()
 
 void mtk_pll_init(int skip_dcm_setting)
 {
-
-	/* Power on PLL */
+	uint32_t readreg;
+	mmio_clrbits_32(ARMPLL_CON1, ARMPLL_CON1_ALL0); // clear all armpll_con1
+	readreg = mmio_read_32(ARMPLL_CON1); //read con1
+	NOTICE("CON1 should 0 actual %X",readreg); //print con1
+	mmio_write_32(ARMPLL_CON1, 0x52000000);
+	mmio_clrbits_32(ARMPLL_CON0 , DIVIDE_RATIO_BIT4);
+
+
 	mmio_setbits_32(ARMPLL_PWR_CON0, CON0_PWR_ON);
 	mmio_setbits_32(NET2PLL_PWR_CON0, CON0_PWR_ON);
 	mmio_setbits_32(MMPLL_PWR_CON0, CON0_PWR_ON);
@@ -92,13 +98,18 @@ void mtk_pll_init(int skip_dcm_setting)
 	mmio_clrbits_32(NET1PLL1_PWR_CON0, CON0_ISO_EN);
 	mmio_clrbits_32(APLL2_PWR_CON0, CON0_ISO_EN);
 	mmio_clrbits_32(MPLL_PWR_CON0, CON0_ISO_EN);
+
+
+	mmio_clrbits_32(ARMPLL_CON0 , DIVIDE_RATIO_BIT4);  //can clear - depand on armpll_con1
 
 	/* Set PLL frequency */
-	mmio_write_32(ARMPLL_CON1, 0x82000000); /* 1.3G */
 
-	//mmio_setbits_32(ARMPLL_CON0, 0x124); /* divider for 650M */
+	readreg = mmio_read_32(ARMPLL_CON1); //read con1
+	NOTICE("CON1 should 0 actual %X",readreg); //print con1
+	mmio_write_32(ARMPLL_CON1, 0x52000000);
+	mmio_setbits_32(ARMPLL_CON0, 0x104);
 
-	mmio_setbits_32(ARMPLL_CON0, 0x114); /* divider for 1.3G */
+
 
 	mmio_setbits_32(NET2PLL_CON0, 0x114);
 	mmio_setbits_32(MMPLL_CON0, 0x124);
--- a/atf-20231013-0ea67d76a/plat/mediatek/mt7981/drivers/pll/pll.h
+++ b/atf-20231013-0ea67d76a/plat/mediatek/mt7981/drivers/pll/pll.h
@@ -64,7 +64,8 @@
 #define CON0_POST_DIV(x)	((x) << 1)
 #define CON1_PCW_CHG		BIT(31)
 #define UNIV_48M_EN		BIT(27)
-
+#define DIVIDE_RATIO_BIT4	0x10
+#define ARMPLL_CON1_ALL0        0xFF000000
 #define SCP_AXICK_DCM_DIS_EN	BIT(0)
 #define SCP_AXICK_26M_SEL_EN	BIT(4)
 
--- a/atf-20240117-bacca82a8/plat/mediatek/mt7981/drivers/pll/pll.c
+++ b/atf-20240117-bacca82a8/plat/mediatek/mt7981/drivers/pll/pll.c
@@ -94,11 +94,20 @@ void mtk_pll_init(int skip_dcm_setting)
 	mmio_clrbits_32(MPLL_PWR_CON0, CON0_ISO_EN);
 
 	/* Set PLL frequency */
-	mmio_write_32(ARMPLL_CON1, 0x82000000); /* 1.3G */
-
-	//mmio_setbits_32(ARMPLL_CON0, 0x124); /* divider for 650M */
-
-	mmio_setbits_32(ARMPLL_CON0, 0x114); /* divider for 1.3G */
+	// 251218: overclock hack
+	NOTICE("THIS IS AN OVERCLOKING BL2!\n");
+	NOTICE("set arm freq to 1.60 GHz:\n");
+
+	// mmio_write_32(ARMPLL_CON1, 0x82000000); /* 1.3G */
+	mmio_write_32(ARMPLL_CON1, 0x50000000); /* 1.6G */
+	// mmio_write_32(ARMPLL_CON1, 0x55000000); /* 1.7G */
+	NOTICE("ARMPLL_CON1 = 0x%x\n", mmio_read_32(ARMPLL_CON1));
+
+	// mmio_setbits_32(ARMPLL_CON0, 0x124); /* divider for 650M */
+	// mmio_setbits_32(ARMPLL_CON0, 0x114); /* divider for 1.3G */
+	mmio_clrbits_32(ARMPLL_CON0, 0x70); /* divider for >1.3G */
+	mmio_setbits_32(ARMPLL_CON0, 0x104);
+	NOTICE("ARMPLL_CON0 = 0x%x\n", mmio_read_32(ARMPLL_CON0));
 
 	mmio_setbits_32(NET2PLL_CON0, 0x114);
 	mmio_setbits_32(MMPLL_CON0, 0x124);
