--- a/uboot-mtk-20220606/board/mediatek/common/ubi_helper.c
+++ b/uboot-mtk-20220606/board/mediatek/common/ubi_helper.c
@@ -532,9 +532,9 @@ static int boot_from_ubi(struct mtd_info *mtd)
 	if (ret)
 		return ret;
 
-	ret = read_ubi_volume("kernel", (void *)data_load_addr, 0);
+	ret = read_ubi_volume("fit", (void *)data_load_addr, 0);
 	if (ret == -ENODEV)
-		ret = read_ubi_volume("fit", (void *)data_load_addr, 0);
+		ret = read_ubi_volume("kernel", (void *)data_load_addr, 0);
 	if (ret)
 		return ret;
 
--- a/uboot-mtk-20220606/board/mediatek/mt7622/bootmenu_emmc.c
+++ b/uboot-mtk-20220606/board/mediatek/mt7622/bootmenu_emmc.c
@@ -291,9 +291,9 @@ int board_boot_default(void)
 #else
 	int ret;
 
-	ret = boot_from_mmc_partition(EMMC_DEV_INDEX, 0, PART_KERNEL_NAME);
+	ret = boot_from_mmc_partition(EMMC_DEV_INDEX, 0, PART_PRODUCTION_NAME);
 	if (ret == -ENODEV)
-		return boot_from_mmc_partition(EMMC_DEV_INDEX, 0, PART_PRODUCTION_NAME);
+		return boot_from_mmc_partition(EMMC_DEV_INDEX, 0, PART_KERNEL_NAME);
 
 	return ret;
 #endif /* CONFIG_MTK_DUAL_BOOT */
--- a/uboot-mtk-20220606/board/mediatek/mt7981/bootmenu_emmc.c
+++ b/uboot-mtk-20220606/board/mediatek/mt7981/bootmenu_emmc.c
@@ -291,9 +291,9 @@ int board_boot_default(void)
 #else
 	int ret;
 
-	ret = boot_from_mmc_partition(EMMC_DEV_INDEX, 0, PART_KERNEL_NAME);
+	ret = boot_from_mmc_partition(EMMC_DEV_INDEX, 0, PART_PRODUCTION_NAME);
 	if (ret == -ENODEV)
-		return boot_from_mmc_partition(EMMC_DEV_INDEX, 0, PART_PRODUCTION_NAME);
+		return boot_from_mmc_partition(EMMC_DEV_INDEX, 0, PART_KERNEL_NAME);
 
 	return ret;
 #endif /* CONFIG_MTK_DUAL_BOOT */
--- a/uboot-mtk-20220606/board/mediatek/mt7986/bootmenu_emmc.c
+++ b/uboot-mtk-20220606/board/mediatek/mt7986/bootmenu_emmc.c
@@ -291,9 +291,9 @@ int board_boot_default(void)
 #else
 	int ret;
 
-	ret = boot_from_mmc_partition(EMMC_DEV_INDEX, 0, PART_KERNEL_NAME);
+	ret = boot_from_mmc_partition(EMMC_DEV_INDEX, 0, PART_PRODUCTION_NAME);
 	if (ret == -ENODEV)
-		return boot_from_mmc_partition(EMMC_DEV_INDEX, 0, PART_PRODUCTION_NAME);
+		return boot_from_mmc_partition(EMMC_DEV_INDEX, 0, PART_KERNEL_NAME);
 
 	return ret;
 #endif /* CONFIG_MTK_DUAL_BOOT */
--- a/uboot-mtk-20220606/board/mediatek/mt7988/bootmenu_emmc.c
+++ b/uboot-mtk-20220606/board/mediatek/mt7988/bootmenu_emmc.c
@@ -291,9 +291,9 @@ int board_boot_default(void)
 #else
 	int ret;
 
-	ret = boot_from_mmc_partition(EMMC_DEV_INDEX, 0, PART_KERNEL_NAME);
+	ret = boot_from_mmc_partition(EMMC_DEV_INDEX, 0, PART_PRODUCTION_NAME);
 	if (ret == -ENODEV)
-		return boot_from_mmc_partition(EMMC_DEV_INDEX, 0, PART_PRODUCTION_NAME);
+		return boot_from_mmc_partition(EMMC_DEV_INDEX, 0, PART_KERNEL_NAME);
 
 	return ret;
 #endif /* CONFIG_MTK_DUAL_BOOT */
--- a/uboot-mtk-20220606/configs/mt7986_redmi_ax6000_defconfig
+++ b/uboot-mtk-20220606/configs/mt7986_redmi_ax6000_defconfig
@@ -4,11 +4,8 @@ CONFIG_ARCH_MEDIATEK=y
 CONFIG_SYS_TEXT_BASE=0x41e00000
 CONFIG_SYS_MALLOC_F_LEN=0x4000
 CONFIG_NR_DRAM_BANKS=1
-CONFIG_ENV_SIZE=0x10000
-CONFIG_ENV_OFFSET=0x100000
 CONFIG_DEFAULT_DEVICE_TREE="mt7986a-redmi-ax6000"
 CONFIG_TARGET_MT7986=y
-CONFIG_ENABLE_NAND_NMBM=y
 CONFIG_MEDIATEK_BOOTMENU=y
 CONFIG_MEDIATEK_BOOTMENU_DELAY=3
 CONFIG_MT7986_BOOTMENU_UBI=y
@@ -39,7 +36,6 @@ CONFIG_CMD_GPIO=y
 CONFIG_CMD_GPIO_READ=y
 CONFIG_CMD_MMC=y
 CONFIG_CMD_MTD=y
-CONFIG_CMD_NMBM=y
 CONFIG_CMD_PCI=y
 CONFIG_CMD_TFTPPUT=y
 # CONFIG_CMD_NFS is not set
@@ -47,8 +43,8 @@ CONFIG_CMD_PING=y
 CONFIG_CMD_LED_BLINK=y
 CONFIG_CMD_SMC=y
 CONFIG_CMD_MTDPARTS=y
-CONFIG_MTDIDS_DEFAULT="nmbm0=nmbm0"
-CONFIG_MTDPARTS_DEFAULT="nmbm0:1024k(bl2),256k(Nvram),256k(Bdata),2048k(factory),2048k(fip),256k(crash),256k(crash_log),112640k(ubi)"
+CONFIG_MTDIDS_DEFAULT="spi-nand0=spi-nand0"
+CONFIG_MTDPARTS_DEFAULT="spi-nand0:1024k(bl2),256k(Nvram),256k(Bdata),2048k(factory),2048k(fip),125440k(ubi)"
 CONFIG_CMD_UBI=y
 CONFIG_CMD_GL_BTN=y
 CONFIG_DOS_PARTITION=y
@@ -56,11 +52,13 @@ CONFIG_EFI_PARTITION=y
 CONFIG_PARTITION_TYPE_GUID=y
 CONFIG_OF_EMBED=y
 CONFIG_ENV_OVERWRITE=y
-CONFIG_ENV_IS_IN_MTD=y
-CONFIG_ENV_MTD_NAME="nmbm0"
-CONFIG_ENV_SIZE_REDUND=0x80000
+CONFIG_ENV_IS_IN_UBI=y
+CONFIG_SYS_REDUNDAND_ENVIRONMENT=y
+CONFIG_ENV_UBI_PART="ubi"
+CONFIG_ENV_UBI_VOLUME="ubootenv"
+CONFIG_ENV_UBI_VOLUME_REDUND="ubootenv2"
 CONFIG_ENV_VARS_UBOOT_RUNTIME_CONFIG=y
 CONFIG_NET_RANDOM_ETHADDR=y
 CONFIG_NET_FORCE_IPADDR=y
 CONFIG_REGMAP=y
 CONFIG_SYSCON=y
 CONFIG_BUTTON=y
--- a/atf-20220606-637ba581b/configs/mt7986_redmi_ax6000_defconfig
+++ b/atf-20220606-637ba581b/configs/mt7986_redmi_ax6000_defconfig
@@ -2,3 +2,4 @@ CONFIG_PLAT_MT7986=y
 CONFIG_TARGET_FIP_NO_SEC_BOOT=y
 CONFIG_FLASH_DEVICE_SPIM_NAND=y
 CONFIG_DRAM_DDR4=y
+# CONFIG_NMBM is not set
